config/config.go

```diff
@@ -22,6 +22,10 @@ type Config struct {
 	Routing   Routing   // local node's routing settings
 	Ipns      Ipns      // Ipns settings
 	Bootstrap []string  // local nodes's bootstrap peer addresses
+	// FIXME(BLOCKING): Review name and documentation.
+	// List of connected peers saved to be used as temporary bootstrap nodes
+	// in case we fail to connect to the official Bootstrap list.
+	TempBootstrapPeers []string
```

Comment:
Sorry if this is discussed somewhere and I missed it, but why can't we use a persistent peerstore? Or could we shove this in a record in the top-level datastore? It makes me uncomfortable storing runtime state in the config like this.

Reply:
Also is there a reason we don't want to try these *in addition to* the bootstrap nodes, or do this the other way around (try these first, then fallback to bootstrappers)?

Reply:
> Sorry if this is discussed somewhere and I missed it, but why can't we use a persistent peerstore? Or could we shove this in a record in the top-level datastore? It makes me uncomfortable storing runtime state in the config like this.

No reason. This was underspecified in the original issue AFAICT so I went with the simplest and most straightforward way to implement and test this. Feel free to discuss in the issue and once decided request the change here.

Reply:
> Also is there a reason we don't want to try these in addition to the bootstrap nodes, or do this the other way around (try these first, then fallback to bootstrappers)?

Not mutually exclusive. Take a look at the `bootstrapRound` implementation:

https://github.com/ipfs/go-ipfs/pull/8856/files#diff-006e79be60d4825ac32ec1713b612812059cb3e84f054efc7624d574e2e97717R217

But if the documentation is misleading on this we can reword it however you like.

Reply:
> No reason. This was underspecified in the original issue AFAICT so I went with the simplest and most straightforward way to implement and test this. Feel free to discuss in the issue and once decided request the change here.

Also see original issue discussion in https://github.com/ipfs/go-ipfs/issues/3926#issuecomment-1067246366 where it seems work on processing and storing this at the libp2p level (which is where this should happen, I agree) has been stalled. This PR is going with the MVP proposal which is not the optimal one.

Reply:
> It makes me uncomfortable storing runtime state in the config like this.

Thinking a bit more about this: I do not see this information as runtime _state_. It is dynamic in the sense it's collected and recycled periodically from a running daemon but it does not carry state information. You do not need this to operate the daemon, it's a fallback and one the user can easily manipulate (expand, backup, reduce) if it wants to select/preserve some peers over others (since this is a very naive implementation that select peers randomly without accessing libp2p-level information).

Reply:
The config continuously changes over time even though the user didn't explicitly change anything, which is surprising for config. I think there are multiple benefits to using the root datastore--it's less surprising (config remains static), and it doesn't expose the state directly to users which gives us more implementation flexibility.

Reply:
> Not mutually exclusive. Take a look at the bootstrapRound implementation:

Ah thanks, that's great. Maybe nice to guarantee some mix of well-known bootstrappers + hysteretic bootstrappers, but we can always tweak that over time.

Reply:
> even though the user didn't explicitly change anything, which is surprising for config.

Fair enough. The config is something that should be explicitly written by the user and not implicitly by the daemon.

> I think there are multiple benefits to using the root datastore

I'm fine with this but need it stated more explicitly if I have to make a substantial change like this one. I have trouble sometimes  differentiating between ongoing discussions brainstorming possible directions and actual expressions of concrete changes needed to land a feature. That's the rationale behind my push to move this thread to the original issue if you're sourcing for opinions or make a review request signaling what needs to change.

Reply:
> mix of well-known bootstrappers + hysteretic bootstrappers,

I'm not sure what a hysteretic bootstrapper is but sounds like the kind of information that needs to be sourced from the libp2p layer (in terms of peers score) and not here.

Reply:
> safe play here is to try static and dynamic bootstraps independently, but in parallel

https://github.com/ipfs/go-ipfs/pull/8856#discussion_r846072335

Reply:
> ephemeral peers do not belong to the config file

ACKed in comment above, but I'm still pushing to move this discussion outside this thread that is already becoming too long to be parsed by others.

Reply:
> storing this at the libp2p level (which is where this should happen, I agree) has been stalled

@schomatis 
FYI, ipfs-cluster do that, it has a text backed peerstore and use that for bootstraping:

https://github.com/ipfs/ipfs-cluster/blob/1707c43b8fb1e6e9063ffb50b5335660c570325b/pstoremgr/pstoremgr.go#L295-L336

Reply:
> but maybe a flat plain text file at $IPFS_PATH/recent-peers (one multiaddr per line) will bring more utility

That exactly what ipfs-cluster do already, it seems to like we should lift that code from ipfs-cluster into a general `libp2p/...` thing that we can import.

Reply:
> we could move them to peerstore/datastore, but maybe a flat plain text file at $IPFS_PATH/recent-peers (one multiaddr per line) will bring more utility? this way people could share this file with each other, we could even have a website which crawls DHT and gives people N peers for download

Better than config. We could add a command with effectively the same interface as what you're talking about, but witthout exposing the persistence details. Since we know this isn't the ideal implementation, it'd be nice to make it easier on ourselves later to change (e.g. move to libp2p).